Proposed changes for issuer parameter

alexkeramidas opened the pull request:
- Removed hardcoded issuer url since most of the time is equal to the site option in client options
- In case it under a path (e.g. `https://dev.connect.authentiq.io/backchannel-logout/`) it is declared in the client options like site, token_url, authorize_url with the parameter jwt_issuer.
```yaml
      client_options: {
          :site => 'https://dev.connect.authentiq.io/',
          :authorize_url => '/backchannel-logout/authorize',
          :token_url => '/backchannel-logout/token',
          :jwt_issuer => 'https://dev.connect.authentiq.io/backchannel-logout/'
      }
```
- Bumped version
- Removed `enable_remote_sign_out` from options since now it will be enabled out of the box and notify the implementor if something is going wrong (he forgot to implement the lambda for example)

alexkeramidas committed:
Proposed changes for issuer

skion commented:
- Could we rename `client_options.site` to `client_options.provider_url`?
- Could we then remove `jwt_issuer`, and just use `provider_url` instead?
- Perhaps rename `BASE_URL` to `DEFAULT_PROVIDER_URL` to indicate its a default.
- What happens if someone places an absolute URL in either `authorize_url` or `token_url`? The extension `_url` would suggest that will work, and will not prepend the `base_url` in that case. Or should we find alternative names for these, to indicate they must be paths.
- Could we remove the `client_options` bit and place all options on the root, since the options in there are not really client/RP configs but more module or server configs.
- How do the `options` in this module relate to the options in the [GitLab config file](https://gitlab.com/alexkeramidas/gitlab-ce/commit/6e186b76bb22c5803e3d7105f8ff78c11128812d#0ef14ca7e447f477230136ca76fb4ab0bed43af4)? Is it expected that they seem entirely different?

alexkeramidas commented:
_- Could we rename client_options.site to client_options.provider_url?_
No since the base omniauth gem relies on that name.

_- Could we then remove jwt_issuer, and just use provider_url instead?_
No, the site parameter needs to be a domain with no ending slash.

_- Perhaps rename BASE_URL to DEFAULT_PROVIDER_URL to indicate its a default._
No problem in changing that to whatever we like. It isnt even needed actually.

_- What happens if someone places an absolute URL in either authorize_url or token_url? The extension _url would suggest that will work, and will not prepend the base_url in that case. Or should we find alternative names for these, to indicate they must be paths._
Yes it works with absolute URLs but the names cannot change since they are declared by (and used from) the base omniauth gem. E.g Github gem uses absolute, auth0 gem uses relative, google uses relative etc. So its up to us.

_- Could we remove the client_options bit and place all options on the root, since the options in there are not really client/RP configs but more module or server configs._
Since there is a a hash named options.client_options in the base omniauth gem, to override something in it, you need to write it as a hash with the same structure and naming. So probably not.

 _-How do the options in this module relate to the options in the GitLab config file? Is it expected that they seem entirely different?_
They are the same actually. 
The options object contains these in the base Oauth2 gem
option :client_id, nil
option :client_secret, nil
option :client_options, {}
option :authorize_params, {}
option :authorize_options, [:scope]
option :token_params, {}
option :token_options, []
option :auth_token_params, {}
option :provider_ignores_state, false

and we overrride it accordingly. Maybe we can remove the args part from the gitlab config file but personally I would not steer that far away from the way the other providers write their config.

Actually, I can remove this line https://github.com/AuthentiqID/omniauth-authentiq/blob/master/lib/omniauth/strategies/authentiq.rb#L18 since its already defined on the base omniauth gem.

skion commented:
>> Could we then remove jwt_issuer, and just use provider_url instead?

> No, the site parameter needs to be a domain with no ending slash.

That actually sounds like a bug in Omniauth, since it should be possible to host an OP on any path, not just the root. The [OIDC Discovery spec](https://openid.net/specs/openid-connect-discovery-1_0.html#IssuerDiscovery) writes: 

> The returned Issuer location MUST be a URI RFC 3986 with a scheme component that MUST be https, a host component, and optionally, port and path components and no query or fragment components.

If OmniAuth would support that correctly, then we would could throw out the `jwt_issuer` option (since by definition it is the `site` URL then) and also not need to include the base path in the `authorize_uri` and `token_uri` any longer.

We could start that in parallel maybe and until then keep the `jwt_issuer` (or better `issuer` since named like that in [OIDC MetaData](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata)) undocumented?

alexkeramidas commented:
Hm now that you mention it, it actually is not mentioned anywhere that omniauth goes by any specific spec.

And yes you are correct, we could throw out the jwt_issuer (or better issuer url) which is actually needed only when the endpoints are not attached to a base url but a path. Otherwise it would only need the site parameter to work.

Does start in parallel mean: Make a PR to omniauth to change their handling of the site parameter?

skion commented:
In parallel: Yes, to submit a PR that allows for a path component in the `site` option. But until that time use `issuer_uri` to override `site` as the expected issuer claim.

ziogaschr commented:
Maybe `omniauth_oauth2` follows the OAuth2 spec. If so, then we can make a PR there for https://github.com/AuthentiqID/omniauth-authentiq/pull/4#issuecomment-280267637

alexkeramidas commented:
omniauth_oauth2 relies on oauth2 gem that provides these naming conventions.

Little details on what the parameters must be are provided here.
For site it says `:site the OAuth2 provider site host`
https://github.com/intridea/oauth2/blob/master/lib/oauth2/client.rb#L11

Since this is using the [Faraday](https://github.com/lostisland/faraday) gem which is an Http client library the site parameter depends on Faraday handling. Which is handled [here](https://github.com/intridea/oauth2/blob/master/lib/oauth2/client.rb#L57).

So if we are to do a PR, I suspect we should do it in either in the OAuth2 gem of the Intrideo company or in the Faraday gem. Probably the 1st

alexkeramidas committed:
Suggested changes by the team

alexkeramidas commented:
Updated the proposed changes. 
- Moved to absolute urls, 
- Removed issuer function and made issuer JWT claim to match the `@options.client_options.site` parameter.
- Removed redundant lines for `BASE_URL` and `:authorize_options`
- Removed outdated check for the events claim.
- Added more detailed validation of id_token
- Refactored algorithm function to a helper file since it is now used both on login and logout

alexkeramidas committed:
More proposed changes

alexkeramidas committed:
More elaborate responses on backchannel logout

alexkeramidas committed:
Merge branch 'master' into backchannel-logout-support
